primitives/src/balances.rs

```diff
@@ -56,6 +61,89 @@ pub struct BalanceEntry<BlockNumber> {
 	pub last_update: BlockNumber,
 }
 
+impl<BlockNumber> BalanceEntry<BlockNumber>
+where
+	BlockNumber: AtLeast32Bit + Debug,
+{
+	pub fn new(principal: BalanceType, last_update: BlockNumber) -> Self {
+		Self { principal, last_update }
+	}
+
+	/// Applies the demurrage and returns an updated BalanceEntry.
+	///
+	/// The following formula is applied to the principal:
+	///    updated_principal = old_principal * e^(-demurrage_per_block * elapsed_blocks)
+	///
+	/// If the `demurrage_per_block` is negative it will take the absolute value of
+	/// it for the calculation.
+	///
+	/// **Note**: This function will be used at every single transaction that is paid with community
+	/// currency. It is important that it is as efficient as possible, but also that it is bullet-
+	/// proof (no-hidden panics!!).
+	pub fn apply_demurrage(
+		self,
+		demurrage_per_block: Demurrage,
+		current_block_number: BlockNumber,
+	) -> BalanceEntry<BlockNumber> {
+		if self.last_update == current_block_number {
+			// Nothing to be done, as no time elapsed.
+			return self
+		}
+
+		if self.principal.eq(&0i16) {
+			return Self { principal: self.principal, last_update: current_block_number }
+		}
+
+		let elapsed_blocks =
+			current_block_number.checked_sub(&self.last_update).unwrap_or_else(|| {
+				// This should never be the case on a sound blockchain setup, but we really
+				// never want to panic here.
+				warn!(
+					target: LOG,
+					"last block {:?} bigger than current {:?}, defaulting to elapsed_blocks=0",
+					self.last_update,
+					current_block_number
+				);
+				0u32.into()
+			});
+
+		let elapsed_u32: u32 = elapsed_blocks.try_into().unwrap_or_else(|_| {
+			// 698 years with 6 seconds block time.
+			trace!(target: LOG, "elapsed_blocks > u32::MAX, defaulting to u32::MAX");
+			u32::MAX
+		});
+
+		let demurrage_factor = demurrage_factor(demurrage_per_block, elapsed_u32);
+
+		let principal = self
+			.principal
+			.checked_mul(demurrage_factor)
+			.expect("demurrage_factor [0,1), hence can't overflow; qed");
+
+		Self { principal, last_update: current_block_number }
+	}
+}
+
+/// e^(-demurrage_per_block * elapsed_blocks) within [0,1).
+///
+/// It will take the absolute value of the `demurrage_per_block` if it is negative.
+pub fn demurrage_factor(demurrage_per_block: Demurrage, elapsed_blocks: u32) -> BalanceType {
+	// We can only have errors here if one of the operations overflowed.
+	//
+	// However, as we compute exp(-x), which goes to 0 for big x, we can
+	// approximate the result as 0 if we overflow somewhere on the way
+	// because of the big x.
+
+	// demurrage >= 0; hence exponent <= 0
+	let exponent = match (-demurrage_per_block.abs()).checked_mul(elapsed_blocks.into()) {
+		Some(exp) => exp,
+		None => return 0.into(),
```

Comment:
zero would cause exp(0) to be 1 which would be returned as demurrage_factor. However, if checked_mul would overflow (which can't happen with sound demurrage values) we should rather skip exp and return 0 as a result, because elapsed_blocks is very high, so 0 is more reasonable than 1.

Reply:
Good catch! But the return 0 here does really return 0 from the function. So it already is, as you say. :)

Reply:
If this weren't the case, this test would not pass: https://github.com/encointer/pallets/blob/681154a785b70c039fb6a8a9fd24fc1166be2505/primitives/src/balances.rs#L279

Reply:
Ah, misread the code. All good